pmd-java/src/main/java/net/sourceforge/pmd/lang/java/ast/ExprContext.java

```diff
@@ -0,0 +1,215 @@
+/*
+ * BSD-style license; for more info see http://pmd.sourceforge.net/license.html
+ */
+
+package net.sourceforge.pmd.lang.java.ast;
+
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+import net.sourceforge.pmd.annotation.Experimental;
+import net.sourceforge.pmd.lang.java.types.JPrimitiveType;
+import net.sourceforge.pmd.lang.java.types.JTypeMirror;
+import net.sourceforge.pmd.lang.java.types.OverloadSelectionResult;
+
+/**
+ * Context of an expression. This determines the target type of poly
+ * expressions, which is necessary for overload resolution. It also
+ * determines what kinds of conversions apply to the value to make it
+ * compatible with the context.
+ *
+ * <p>The api is minimal until more use cases show up, and this is better
+ * tested.
+ */
+@Experimental
+public abstract class ExprContext {
```

Comment:
This is in the main `java.ast` package, but it deals with typeres. Would this fit better into `java.types`? Maybe along with PolyResolution? Just a thought....

Reply:
If we move them, we'd have to also move LazyTypeResolver and make it public. I think, this would make us expose more members than necessary... Also LazyTypeResolver uses some package-private apis of the ast package. Maybe something to keep in mind, but for now I think this is the state of lowest energy